src/mlpack/methods/ann/layer/scaled_dot_product_attention.hpp

```diff
@@ -0,0 +1,228 @@
+/**
+ * @file methods/ann/layer/scaled_dot_product_attention.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Definition of the ScaledDotProductAttention class.
+ *
+ * @code
+ * @article{NIPS'17,
+ *   author  = {Ashish Vaswani, Llion Jones, Noam Shazeer, Niki Parmar,
+ *              Aidan N. Gomez, Jakob Uszkoreit, Łukasz Kaiser,
+ *              Illia Polosukhin},
+ *   title   = {Attention Is All You Need},
+ *   year    = {2017},
+ *   url     = {http://arxiv.org/abs/1706.03762v5}
+ * }
+ * @endcode
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_SCALED_DOT_PRODUCT_ATTENTION_HPP
+#define MLPACK_METHODS_ANN_LAYER_SCALED_DOT_PRODUCT_ATTENTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "layer_types.hpp"
+#include "softmax.hpp"
+#include "../init_rules/glorot_init.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class ScaledDotProductAttention
+{
+ public:
+  //! Element Type of the input.
+  typedef typename InputDataType::elem_type ElemType;
+
+  /**
+   * Default constructor.
+   */
+  ScaledDotProductAttention();
+
+  /**
+   * Create the ScaledDotProductAttention object using the specified parameters.
+   *
+   * @param embedDim Total dimension of the model.
+   * @param key The key matrix.
+   * @param value The value matrix.
+   * @param dropoutRate The dropout rate for attention output weights.
+   * @param deterministic If false, dropout layer is omitted else dropout layer
+   *        is applied with dropout rate `dropout`.
+   */
+  ScaledDotProductAttention(const size_t embedDim,
+    const InputDataType& key = InputDataType(),
+    const InputDataType& value = InputDataType(),
+    const ElemType dropoutRate = 0.1,
+    const bool deterministic = false);
```

Comment:
```suggestion
  ScaledDotProductAttention(
      const size_t embedDim,
      const InputDataType& key = InputDataType(),
      const InputDataType& value = InputDataType(),
      const ElemType dropoutRate = 0.1,
      const bool deterministic = false);
```

Just to be consistent with the rest of the codebase.